### PR TITLE
feat: download snippet generation

### PR DESCRIPTION
**Pull request by ovflowd**
This PR attempts to introduce a system for dynamic download snippet generation. With the increasing number of different version managers, package managers, and installation methods for Node.js, we want to ease maintainability, translation, and viewing how the snippet would look with syntax highlighting.

This API allows these files to be loaded on demand, and then Shiki and MDX transpile them on the client side. This requires having Shiki and the MDX compiler present on the client side (which they already are).

This change introduces the ability to test the snippets in a sandboxed environment and, in the future, have every CodeBox be loaded on the client-side dynamically (including Learn codeboxes) on-demand without requiring them to be built during build time.

The trade-off is that the client side needs to fetch these snippets, but this request is prefetched and cached.

This introduces the API route, data generators and providers + replaces the usage of the old getDownloadSnippet with the new mechanism.

**This PR also...**

- It makes Shiki compilation a fully synchronous process, removing the need for some asynchronous pieces and top-level awaits and being more client-side friendly. It also reduces ticks as multiple MDX compilation processes will not need to resolve Promises/ticks
- Reduces usage of useDetectOS and optimizes it to use synchronous OS detection and reduce layout shift and state change as much as possible, keeping the initial state as clean as possible
- Update the ReleaseCodeBox so that it can render the CodeBox in a fully synchronous way which is React-friendly and kills the CodeBox loading state / being empty/shifting values
- Updates the Select component in such a way that it removes the flashing/initial state being empty for a while
- Multiple minor bundle optimization and initial rendering improvements related to Downloads pages and overall App
- Fixes Vercel Environment variables and errors on first builds (consuming wrong sources during self-consumption process)
- Fixes Avatar rendering and placement / optimizing the rendering

### CODE REVIEW

File: apps/site/components/JSX/CodeBox/index.tsx [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+type CodeBoxProps = { language: string; showCopyButton?: string };
+
+const codeToHast = highlightToHast(await shikiPromise);
+
+const CodeBox: FC<PropsWithChildren<CodeBoxProps>> = async props => {
+  const out = codeToHast(dedent(props.children as string), props.language);
+  const className = `language-${props.language}`;
+
+  return toJsxRuntime(out, {
```
Comment (ovflowd) [inline apps/site/components/JSX/CodeBox/index.tsx]: I realize this bypassed our MDX Rehype Plugin, unironically. We support three types of CodeBoxes now:

- MDXCodeBox, source comes from MDX code blocks and code snippet will be transpilled within the MDX parsing process.
- CodeBox, plain base CodeBox component without syntax highlighting
- JSXCodeBox, the source is a children string/ReactNode, and the code snippet gets transpilled directly within the component.

I assume the difference between the MDX and JSX is that the MDX requires plain code blocks within a Markdown/MDX file, and the JSX one uses a JSX component within MDX but does the transpilation outside of MDX.

Both use the same memoized Shiki instance.

**Comment by MattIPv4** [comment]
> Note that this (at the time being) prevents the generation of the CodeBox for static exports since the combination of language+node version+installation method would blow the generation away.

This concerns me -- I would think we would absolutely want to prioritise these snippets being in the HTML sent over the wire initially as they're rather crucial content, and likely content that bots etc. would want to index/scrape too.

File: apps/site/app/[locale]/next-data/download-snippets/route.tsx [hunk 2]
```diff
@@ -0,0 +1,37 @@
+import provideDownloadSnippets from '@/next-data/providers/downloadSnippets';
+import { defaultLocale } from '@/next.locales.mjs';
+
+type DynamicStaticPaths = { locale: string };
+type StaticParams = { params: Promise<DynamicStaticPaths> };
+
+// This is the Route Handler for the `GET` method which handles the request
+// for generating OpenGraph images for Blog Posts and Pages
```
Comment (AugustinMauroy) [inline apps/site/app/[locale]/next-data/download-snippets/route.tsx]: This comment isn't true

File: crowdin.yml [hunk 3]
```diff
@@ -32,6 +32,15 @@ files:
         pt-BR: pt-br
         zh-CN: zh-cn
         zh-TW: zh-tw
+  - source: /apps/site/snippets/en/**/*.bash
```
Comment (AugustinMauroy) [inline crowdin.yml:35]: is it dangerous to translate that ?

**Review by AugustinMauroy — DISMISSED** [review]

File: crowdin.yml (hunk 3 above)
Comment (ovflowd) [reply crowdin.yml:35]: These are never executed... Although the extension is .bash they are pretty much just text...

Comment (ovflowd) [reply crowdin.yml:35]: Also they don't have chmod +x by default

Comment (AugustinMauroy) [reply crowdin.yml:35]: I mean dangerous like they can be break in the "past" we just translate the comment not the whole script

Comment (AugustinMauroy) [reply crowdin.yml:35]: so my question is, do we need to translate theses file ?

Comment (ovflowd) [reply crowdin.yml:35]: Yes, we do.

> I mean dangerous like they can be break in the "past" we just translate the comment not the whole script

I don't think it is dangerous, the issue with MDX was that MDX uses a custom engine and has special things. This is plain text, no way for mistakes to happen.

File: apps/site/components/Common/Skeleton/index.module.css [hunk 4]
```diff
@@ -0,0 +1,27 @@
+.skeleton {
+  @apply dark:animate-pulse-dark
```
Comment (canerakdas) [inline apps/site/components/Common/Skeleton/index.module.css:2]:
```suggestion
  @apply motion-safe:dark:animate-pulse-dark
```

File: apps/site/components/Common/Skeleton/index.module.css [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    pointer-events-none
+    animate-pulse
```
Comment (canerakdas) [inline apps/site/components/Common/Skeleton/index.module.css:4]:
```suggestion
    motion-safe:animate-pulse
```

File: apps/site/components/Common/Skeleton/index.module.css [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    shadow-none
+    outline-none;
+}
+
+.skeleton[data-inline-skeleton] {
+  @apply leading-none;
+}
+
+.skeleton:empty {
+  @apply block h-3;
```
Comment (canerakdas) [inline apps/site/components/Common/Skeleton/index.module.css]:
```suggestion
  @apply block
    h-3;
```

**Review by canerakdas — APPROVED** [review]
I added a few minor comments related with CSS, it's great that the avatars are optimized and the server-side dynamic download snippet generation 🤩

I also compared performance and page file sizes, everything looks great ✨

**Comment by ljharb** [comment]
This PR removed the "install nvm" (and "install fnm", etc) links, which is a major disruption to the UX flow here. Can those please be restored, or this PR reverted, ASAP?

**Comment by ovflowd** [comment]
> This PR removed the "install nvm" (and "install fnm", etc) links, which is a major disruption to the UX flow here. Can those please be restored, or this PR reverted, ASAP?

Yes, apologies, this shouldn't have happened. It was a mistake that the updated instructions without the install scripts got merged. As per DM conversation on Slack, feel free to re-add it asap, and I appreciate you for noticing it 🙇

**Comment by ljharb** [comment]
→ https://github.com/nodejs/nodejs.org/pull/7360

**Pull request merged**